src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs

```diff
@@ -974,12 +974,13 @@ private bool IsValidWellKnownType(NamedTypeSymbol? result)
 
         private static NamedTypeSymbol? GetTopLevelTypeByMetadataName(AssemblySymbol assembly, ref MetadataTypeName metadataName, AssemblyIdentity? assemblyOpt)
         {
-            var result = assembly.LookupTopLevelMetadataType(ref metadataName, digThroughForwardedTypes: false);
-            if (!IsAcceptableMatchForGetTypeByMetadataName(result))
+            var result = assembly.LookupTopLevelMetadataType(ref metadataName, digThroughForwardedTypes: false, returnNullForMissing: true);
+            if (result is null)
             {
                 return null;
             }
 
+            Debug.Assert(IsAcceptableMatchForGetTypeByMetadataName(result));
             if (assemblyOpt != null && !assemblyOpt.Equals(assembly.Identity))
             {
```

Comment:
This is the important part.
Here, LookupTopLevelMetadataType could end up allocating `MissingMetadataTypeSymbol.TopLevel`, which is thrown away and `null` is returned instead. So, now we pass in `returnNullForMissing: true` so that the allocation isn't made and `null` is returned directly to us.